internal/signing/channelaccounts/channel_accounts_model.go

```diff
@@ -0,0 +1,150 @@
+	UpdatedAt           time.Time    `db:"updated_at"`
+	CreatedAt           time.Time    `db:"created_at"`
+	LockedAt            sql.NullTime `db:"locked_at"`
+	LockedUntil         sql.NullTime `db:"locked_until"`
+}
+
+type ChannelAccountStore interface {
+	GetIdleChannelAccount(ctx context.Context, lockedUntil time.Duration) (*ChannelAccount, error)
+	Get(ctx context.Context, sqlExec db.SQLExecuter, publicKey string) (*ChannelAccount, error)
+	BatchInsert(ctx context.Context, sqlExec db.SQLExecuter, channelAccounts []*ChannelAccount) error
+	Count(ctx context.Context) (int64, error)
+}
+
+type ChannelAccountModel struct {
+	DB db.ConnectionPool
+}
+
+var _ ChannelAccountStore = (*ChannelAccountModel)(nil)
+
+func (ca *ChannelAccountModel) GetIdleChannelAccount(ctx context.Context, lockedUntil time.Duration) (*ChannelAccount, error) {
+	query := fmt.Sprintf(`
+		UPDATE channel_accounts
+		SET 
+			locked_at = NOW(),
+			locked_until = NOW() + INTERVAL '%d seconds'
+		WHERE public_key = (
+			SELECT
+				public_key
+			FROM channel_accounts
+			WHERE 
+				locked_until IS NULL
+				OR locked_until < NOW()
+			ORDER BY random()
+			LIMIT 1
+			FOR UPDATE SKIP LOCKED
+		)
+		RETURNING *
+	`, int64(lockedUntil.Seconds()))
+
+	var channelAccount ChannelAccount
+	for range ChannelAccountWaitTime {
+		err := ca.DB.GetContext(ctx, &channelAccount, query)
+		if errors.Is(err, sql.ErrNoRows) {
+			log.Ctx(ctx).Warn("All channel accounts are in use. Retry in 1 second.")
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if err != nil {
+			return nil, fmt.Errorf("getting idle channel account: %w", err)
+		}
+
+		return &channelAccount, nil
+	}
+
+	return nil, ErrNoAvailableChannelAccount
+}
+
+func (ca *ChannelAccountModel) Get(ctx context.Context, sqlExec db.SQLExecuter, publicKey string) (*ChannelAccount, error) {
+	const query = `SELECT * FROM channel_accounts WHERE public_key = $1`
+
+	var channelAccount ChannelAccount
+	err := sqlExec.GetContext(ctx, &channelAccount, query, publicKey)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, ErrChannelAccountNotFound
+		}
+		return nil, fmt.Errorf("getting channel account %s: %w", publicKey, err)
+	}
+
+	return &channelAccount, nil
+}
+
+func (ca *ChannelAccountModel) BatchInsert(ctx context.Context, sqlExec db.SQLExecuter, channelAccounts []*ChannelAccount) error {
+	if len(channelAccounts) == 0 {
+		return nil
+	}
+
+	publicKeys := make([]string, len(channelAccounts))
+	encryptedPrivateKeys := make([]string, len(channelAccounts))
+	for i, ca := range channelAccounts {
+		if ca.PublicKey == "" {
+			return fmt.Errorf("public key cannot be empty")
+		}
+		if ca.EncryptedPrivateKey == "" {
+			return fmt.Errorf("private key cannot be empty")
+		}
+
+		publicKeys[i] = ca.PublicKey
+		encryptedPrivateKeys[i] = ca.EncryptedPrivateKey
+	}
+
+	const q = `
+		INSERT INTO 
+			channel_accounts (public_key, encrypted_private_key)
+		SELECT * 
+			FROM UNNEST($1::text[], $2::text[])
```

Comment:
One nice alternative for Postgres is this: https://pkg.go.dev/github.com/lib/pq#hdr-Bulk_imports
But might not be such a good idea because it make the implementation more coupled to one database system.

Reply:
I prefer keeping it the way it is now.